packages/javascript-sdk/src/fr-qr-code/index.ts

```diff
@@ -0,0 +1,86 @@
+import { CallbackType } from '../auth/enums';
+import HiddenValueCallback from '../fr-auth/callbacks/hidden-value-callback';
+import TextOutputCallback from '../fr-auth/callbacks/text-output-callback';
+import FRStep from '../fr-auth/fr-step';
+
+type QRCodeData = {
+  message: string;
+  use: string;
+  uri: string;
+};
+
+/**
+ * @class FRQRCode - A utility class for handling QR Code steps
+ *
+ * Example:
+ *
+ * ```js
+ * const isQRCodeStep = FRQRCode.isQRCodeStep(step);
+ * let qrCodeData;
+ * if (isQRCodeStep) {
+ *   qrCodeData = FRQRCode.getQRCodeData(step);
+ * }
+ * ```
+ */
+abstract class FRQRCode {
+  /**
+   * @method isQRCodeStep - determines if step contains QR Code callbacks
+   * @param {FRStep} step - step object from AM response
+   * @returns {boolean}
+   */
+  public static isQRCodeStep(step: FRStep): boolean {
+    const hiddenValueCb = step.getCallbacksOfType(CallbackType.HiddenValueCallback);
+
+    // QR Codes step should have at least one HiddenValueCallback
+    if (hiddenValueCb.length === 0) {
+      return false;
+    }
+    return !!this.getQRCodeURICb(hiddenValueCb);
+  }
+
+  /**
+   * @method getQRCodeData - gets the necessary information from the QR Code callbacks
+   * @param {FRStep} step - step object from AM response
+   * @returns {QRCodeData}
+   */
+  public static getQRCodeData(step: FRStep): QRCodeData {
+    const hiddenValueCb = step.getCallbacksOfType(CallbackType.HiddenValueCallback);
+
+    // QR Codes step should have at least one HiddenValueCallback
+    if (hiddenValueCb.length === 0) {
+      throw new Error(
+        'QR Code step must contain a HiddenValueCallback. Use `FRQRCode.isQRCodeStep` to guard.',
+      );
+    }
+    const qrCodeURICb = this.getQRCodeURICb(hiddenValueCb) as HiddenValueCallback;
+    const outputValue = qrCodeURICb ? qrCodeURICb.getOutputValue('value') : '';
+    const qrCodeUse =
+      typeof outputValue === 'string' && outputValue.includes('otpauth://') ? 'otp' : 'push';
+
+    const messageCbs = step.getCallbacksOfType(CallbackType.TextOutputCallback);
+    const displayMessageCb = messageCbs.find((cb) => {
+      const textOutputCallback = cb as TextOutputCallback;
+      return textOutputCallback.getMessageType() !== '4';
+    }) as TextOutputCallback | null;
```

Comment:
This is odd to me from a typescript perspective. I don't believe its a flaw of the code but maybe is more significant of a smell in the types elsewhere somehow.

```ts
 find<S extends T>(predicate: (value: T, index: number, obj: T[]) => value is S, thisArg?: any): S | undefined;
    find(predicate: (value: T, index: number, obj: T[]) => unknown, thisArg?: any): T | undefined;
```

My understanding is that the overload here used would be the bottom one, since we are not passing a generic in.

the argument `cb` is typed correctly, as a TextOutputCallback because we are iterating a `TextOutputCallback[]`

Where I get lost (and I only checked this on the github cloud editor), is that the type of `getMessageType()` is evaluating to `any` and I can't figure out why in the few minutes i spent here.

The reason its confusing is because the class correctly types this as string.

if this code was refactored like so:

```ts
 const displayMessageCb = messageCbs.find((cb) => {
     return cb.getMessageType() !== '4';
    }) 
```

I would expect the `T` to be the type of `cb` which would be `TextoutputCallback` but i can't understand why its saying `any`.

![Screenshot 2023-09-25 at 9 19 44 AM](https://github.com/ForgeRock/forgerock-javascript-sdk/assets/18267769/600898ae-797b-4d80-8caa-2ebd0f40a060)

Not to spend a lot of time on this, but i'm curious if we should document this as a potential bug (in our types) and try to replicate it somehow if we have time down the road.

This is non blocking for this PR but i'm curious about why this is occuring as it currently is confusing to me. Maybe i'm misreading or misunderstanding something early today but i'll leave this here in case someone else can reason it.

Reply:
also maybe its just github's editor making me bonkers?